### PR TITLE
helper for preparing URLs for use_course()

### LINKED ISSUE #406: Helper for constructing links suitable for use_course()
Provide more help (and a convenient place of docs) re: how to get "download" links for GitHub, DropBox, Google Drive.

If possible, also a way to make short links from usethis.

### PR DESCRIPTION
**Pull request by fmichonneau**
Closes #406

### CODE REVIEW

File: R/course.R [hunk 1]
```diff
@@ -154,6 +155,27 @@ use_zip <- function(url,
 #' https://api.github.com/repos/r-lib/usethis/zipball/master
 #' ```
 #'
+#' **Google Drive:**
+#'
+#' It looks like it's not possible to directly share as a ZIP file the content
+#' of a folder storred in Google Drive. It is however possible to share a ZIP
```
Comment (jennybc) [inline R/course.R]: Bummer. I really thought this was possible, but from my own searching, I agree there doesn't seem to be a workflow that is comparable to what we have for DropBox or GitHub.

File: R/course.R [hunk 2]
```diff
@@ -154,6 +155,27 @@ use_zip <- function(url,
...
+#' file storred in Google Drive. To get this URL, click on
```
Comment (jennybc) [inline R/course.R]:
```suggestion
#' file stored in Google Drive. To get this URL, click on
```

File: R/course.R [hunk 3]
```diff
@@ -154,6 +155,27 @@ use_zip <- function(url,
...
+#' "Get the shareable link" (within the "Share" menu). This URL doesn't allow
+#' for direct download as it's designed to be processed in a web browser first.
```
Comment (jennybc) [inline R/course.R]:
```suggestion
#' for direct download, as it's designed to be processed in a web browser first.
```

File: R/course.R [hunk 4]
```diff
@@ -154,6 +155,27 @@ use_zip <- function(url,
...
+#' Such a sharing link looks like:
+#'
+#' ```
+#' https://drive.google.com/open?id=123456789xxyyyzzz
+#' ```
+#'
+#' To be able to get the URL suitable for direct download, you need to extract
+#' the "id" element from the URL and include it in this URL format (or):
```
Comment (jennybc) [inline R/course.R]: What's up with the dangling "(or)"? Did you mean to mention the new URL-processing function?

File: R/course.R [hunk 5]
```diff
@@ -273,6 +295,25 @@ tidy_unzip <- function(zipfile, cleanup = FALSE) {
   invisible(target)
 }
 
+## takes a Dropbox or Google Drive URL as provided in the web browser, and
```
Comment (jennybc) [inline R/course.R]:
```suggestion
## takes a Dropbox or Google Drive URL as captured from the web browser, and
```

File: R/course.R [hunk 6]
```diff
@@ -273,6 +295,25 @@ tidy_unzip <- function(zipfile, cleanup = FALSE) {
...
+## converts into an URL suitable for direct download (as expected by
+## tidy_download() for instance).
+convert_teaching_url <- function(url) {
+  stopifnot(is.character(url))
```
Comment (jennybc) [inline R/course.R]: If you're really expecting a character of length one, we have `is_string()` for this.

File: R/course.R [hunk 7]
```diff
@@ -273,6 +295,25 @@ tidy_unzip <- function(zipfile, cleanup = FALSE) {
...
+
+  if (grepl("drive.google.com", url)) {
+    id <- regmatches(url, regexpr("id=[^&]+&?", url))
+    return(glue::glue("https://drive.google.com/uc?export=download&{id}"))
+  }
+
+  if (grepl("dropbox.com/sh", url)) {
+    return(gsub("dl=0", "dl=1", url))
+  }
+
```
Comment (jennybc) [inline R/course.R]: I think we should probably use `ui_info()` here to say something like: "Not a recognized URL format. No changes made."

**Comment by hadley** [comment]
@fmichonneau are you interested in finishing this off?

**Comment by fmichonneau** [comment]
yes, I'm interested in finishing this. I'll get to it in the next few days. Thanks for the reminder.

**Comment by hadley** [comment]
@jennybc I think this is good to merge, but needs final approval from you.

**Comment by hadley** [comment]
I think this fixes #406?

**Comment by jennybc** [comment]
Thanks!

**Comment by fmichonneau** [comment]
thank you both for polishing and merging this!

**Pull request merged**